controller/src/main/java/org/jboss/as/controller/management/ManagementInterfaceAddStepHandler.java

```diff
@@ -0,0 +1,97 @@
+import org.jboss.msc.service.ServiceController.State;
+import org.jboss.msc.service.ServiceName;
+import org.jboss.msc.service.ServiceNotFoundException;
+import org.jboss.msc.service.ServiceRegistry;
+import org.wildfly.security.manager.WildFlySecurityManager;
+
+/**
+ * The base add handler for management interfaces.
+ * @author Emmanuel Hugonnet (c) 2016 Red Hat, inc.
+ */
+public abstract class ManagementInterfaceAddStepHandler extends AbstractAddStepHandler {
+
+    protected ManagementInterfaceAddStepHandler(final RuntimeCapability capability, final AttributeDefinition[] attributeDefinitions) {
+        super(capability, attributeDefinitions);
+    }
+
+   @Override
+    protected boolean requiresRuntime(OperationContext context) {
+        //TODO Gigantic HACK to disable the runtime part of this for the core model testing.
+        //The core model testing currently uses RunningMode.ADMIN_ONLY, but in the real world
+        //the http interface needs to be enabled even when that happens.
+        //I don't want to wire up all the services unless I can avoid it, so for now the tests set this system property
+        return WildFlySecurityManager.getPropertyPrivileged("jboss.as.test.disable.runtime", null) == null;
+    }
+
+    protected void addVerifyInstallationStep(OperationContext context, List<ServiceName> requiredServices) {
+        if(context.isBooting()) {
+            context.addStep(new VerifyInstallationStep(requiredServices), OperationContext.Stage.VERIFY);
+        }
+    }
+
+    protected String asStringIfDefined(OperationContext context, AttributeDefinition attribute, ModelNode model) throws OperationFailedException {
+        ModelNode attributeValue = attribute.resolveModelAttribute(context, model);
+        return attributeValue.isDefined() ? attributeValue.asString() : null;
+    }
+
+    private static class VerifyInstallationStep implements OperationStepHandler {
+        private final List<ServiceName> requiredServices;
+
+        private VerifyInstallationStep(List<ServiceName> requiredServices) {
+            assert requiredServices != null;
+            this.requiredServices = requiredServices;
+        }
+
+        @Override
+        public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
+            List<ServiceName> failures = new ArrayList<>();
+            ServiceRegistry registry = context.getServiceRegistry(false);
+            for (ServiceName serviceName : requiredServices) {
+                try {
+                    ServiceController<?> controller = registry.getService(serviceName);
+                    if (controller == null || State.UP != controller.getState()) {
+                        failures.add(serviceName);
+                    }
+                } catch (ServiceNotFoundException ex) {
+                    failures.add(serviceName);
+                }
+            }
+            if (failures.containsAll(requiredServices)) {
```

Comment:
Ah, now I get where ManagementRemotingServices.MANAGEMENT_ENDPOINT was coming from in the HTTP cases. I think you're going for an "either-or" thing where you check all possible services and if any are ok, it's ok.

ManagementRemotingServices.MANAGEMENT_ENDPOINT doesn't suffice though, as the endpoint being up doesn't mean the socket to receive native is up (for native) or the service to do HTTP Upgrade is up (for http.)

I think the required services should be those known to be required by a given handler subclass and no others.  The subclasses should not need to know the details of other subclasses, as that will be fragile.

To be lenient and not fail so long as 1 subclass succeeds, without requiring the various subclasses to know each others' details:

1) ManagementInterfaceAddStepHandler can define a constant OperationContext.AttachmentKey, named for example REMOTE_MANAGEMENT.
2) VerifyInstallationStep checks all the required services, which are only the ones relevant to the given subclass. 
a) If all are ok, it attaches REMOTE_MANAGEMENT to the context.
b) If any are not ok, it doesn't call context.setRollbackOnly() immediately. Instead it adds another VERIFY step. That step checks for the present of context attachment REMOTE_MANAGEMENT. If present, some other handler succeeded so all is well and it just returns. If not present, then none have succeeded and it calls context.setRollbackOnly().